src/parser/core/modules/CooldownDowntime.tsx

```diff
@@ -0,0 +1,264 @@
+	static title = t('core.cooldowndowntime.title')`Cooldown Downtime`
+
+	@dependency private downtime!: Downtime
+	@dependency private checklist!: Checklist
+
+	/**
+	 * Implementing modules MUST provide a list of tracked cooldowns
+	 */
+	protected abstract trackedCds: CooldownGroup[]
+
+	private usages = new Map<number, CastEvent[]>()
+	private resets = new Map<number, CastEvent[]>()
+
+	protected checklistName = <Trans id="core.cooldowndowntime.use-ogcd-cds">Use your OGCDs</Trans>
+	protected checklistDescription = <Trans id="core.cooldowndowntime.ogcd-cd-metric">Always make sure to use your OGCDs when they are up but don't clip them.</Trans>
+	protected checklistTarget = 95
+
+	protected defaultAllowedDowntime = 2500
+	protected defaultFirstUseOffset = 0
+
+	/**
+	 * Implementing modules MAY filter out some usages as 'fake' usages
+	 * of a cooldown.
+	 *
+	 * The primary example of this would be NIN mudras, where any mudra
+	 * may start a usage, but further mudras while the Mudra buff or any
+	 * during the Kassatsu buff should not be counted as usages, even
+	 * though all 3 mudra skills are part of the group.
+	 *
+	 * @param event The cast that may or may not be counted.
+	 * @returns True if the event should be counted or false if the event
+	 * should not be counted as a usage of the cooldown.
+	 */
+	protected countUsage(event: CastEvent): boolean {
+		return true
+	}
+
+	protected init() {
+		const trackedIds = this.trackedCds.map(g => g.cooldowns)
+			.reduce((acc, cur) => acc.concat(cur))
+			.map(a => a.id)
+
+		const resetIds = this.trackedCds
+			.map(g => (g.resetBy === undefined) ? [] : g.resetBy.actions)
+			.reduce((acc, cur) => acc.concat(cur))
+			.map(a => a.id)
+
+		this.addEventHook('cast', {by: 'player', abilityId: trackedIds}, this.onTrackedCast)
+		this.addEventHook('cast', {by: 'player', abilityId: resetIds}, this.onResetCast)
+		this.addEventHook('complete', this.onComplete)
+
+		this.trackedCds.forEach(g => {
+			this.usages.set(g.cooldowns[0].id, [])
+			this.resets.set(g.cooldowns[0].id, [])
+		})
+	}
+
+	private onTrackedCast(event: CastEvent) {
+		if (!this.countUsage(event)) { return }
+
+		const groupId = this.getTrackedGroupId(event.ability.guid)
+		if (groupId === undefined) { return }
+
+		(this.usages.get(groupId) || []).push(event)
+	}
+
+	private getTrackedGroupId(abilityId: number): number | undefined {
+		const group = this.trackedCds.find(g => g.cooldowns.find(a => a.id === abilityId) !== undefined)
+		if (group === undefined) { return undefined }
+		return group.cooldowns[0].id
+	}
+
+	private onResetCast(event: CastEvent) {
+		this.trackedCds.forEach(g => {
+			if (g.resetBy && g.resetBy.actions.find(a => a.id === event.ability.guid)) {
+				(this.resets.get(g.cooldowns[0].id) || [] ).push(event)
+			}
+		})
+	}
+
+	private onComplete() {
+		const cdRequirements = []
+		for (const cdGroup of this.trackedCds) {
+			const expected = this.calculateMaxUsages(cdGroup)
```

Comment:
So; correct me if I'm wrong here - the "forward look" mentioned in the PR description is being performed post-mortem? If so, I have no issue with it - the parse is already complete, so the "forward look" is more like a "backward look but not as backward as the last bit", which is more than fine.

Reply:
Yeah, after the complete flag fires, we start at time 0 (or firstUseOffset) and then crawl forward through the fight to determine when the skill "should" have been used, to count those max usages.